### PR TITLE
CRM-20569: Fix membership on partially paid payment via backoffice form and other fixes

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20569: Record Payment for partially paid membership: update amt, fix deferred ](https://issues.civicrm.org/jira/browse/CRM-20569)

### CODE REVIEW

**Comment by monishdeb** [comment]
Marked as [WIP] as I am writing the unit test for it.

**Comment by eileenmcnaughton** [comment]
All this looks like to me like it might be time to migrate that form to using the payment api?

**Comment by monishdeb** [comment]
@eileenmcnaughton can't :( as it's a time-bound IIDA issue for which I can't put efforts on any refactoring/improvement. However, what if we raise a backlog CRM ticket for the same? About using payment API for backoffice membership and other refactoring changes!

File: CRM/Price/BAO/PriceSet.php [hunk 1]
```diff
@@ -703,8 +703,8 @@ public static function processAmount($fields, &$params, &$lineItem, $component =
     }
     if ($priceSetID) {
       $priceFields = self::filterPriceFieldsFromParams($priceSetID, $params);
-      if (count($priceFields) == 1 && !empty($params['total_amount'])) {
-        $amount_override = $params['total_amount'];
+      if (count($priceFields) == 1) {
+        $amount_override = CRM_Utils_Array::value('partial_payment_total', $params, CRM_Utils_Array::value('total_amount', $params));
```
Comment (monishdeb) [inline CRM/Price/BAO/PriceSet.php]: like https://github.com/civicrm/civicrm-core/blob/master/CRM/Price/BAO/LineItem.php#L558

**Comment by monishdeb** [comment]
The PR fixes four issues:
- Handle partial payment for backoffice membership
- Add formRule validation to avoid partial payment for more the one mbmership Types via backoffice
- Record proper financial record ID related to deferred revenue
- Upon Completing a related Contribution after additional payment, change the desired membership status, calculated on basis of membership start/end/join dates

**Comment by monishdeb** [comment]
@eileenmcnaughton I have added a unit test and did some additional fixes. Please have a look!

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -3844,6 +3844,15 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
       $participantId = CRM_Core_DAO::getFieldValue('CRM_Event_DAO_ParticipantPayment', $contributionId, 'participant_id', 'contribution_id');
     }
 
+    // fetch membership IDs registered aginst given contribution ID
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: s/aginst/against/

File: CRM/Financial/BAO/FinancialItem.php [hunk 3]
```diff
@@ -109,7 +109,10 @@ public static function add($lineItem, $contribution, $taxTrxnID = FALSE, $trxnId
     }
     else {
       $accountRelName = 'Income Account is';
-      if (property_exists($contribution, 'revenue_recognition_date') && !CRM_Utils_System::isNull($contribution->revenue_recognition_date)) {
+      if (($itemStatus == array_search('Unpaid', $financialItemStatus)) &&
```
Comment (JoeMurray) [inline CRM/Financial/BAO/FinancialItem.php]: Just raising a uninformed question in case it hasn't been tested: when reversing out deferred revenue items, is this code used, and would it be appropriate to provide the revenue recognition date and deferred revenue account?

File: CRM/Member/BAO/Membership.php [hunk 4]
```diff
@@ -2398,6 +2398,8 @@ public static function recordMembershipContribution(&$params, $ids = array()) {
       'contribution_recur_id',
       'pan_truncation',
       'card_type_id',
+      'partial_payment_total',
+      'partial_amount_pay',
```
Comment (JoeMurray) [inline CRM/Member/BAO/Membership.php]: Just wondering about name: should it be partial_amount_paid if it represents what is being paid now? If it represents what remains to be paid, how about partial_amount_to_pay ?

File: CRM/Member/Form/Membership.php [hunk 5]
```diff
@@ -122,16 +122,17 @@ class CRM_Member_Form_Membership extends CRM_Member_Form {
...
-          $memTypeSelected[$id] = $id;
-        }
+
+    foreach ($priceFieldIDS as $priceFieldId) {
+      if ($memTypeID = CRM_Core_DAO::getFieldValue('CRM_Price_DAO_PriceFieldValue', $priceFieldId, 'membership_type_id')) {
+        $memTypeSelected[$priceFieldId] = $memTypeID;
       }
     }
+
+    if (empty($memTypeSelected) && isset($params['membership_type_id']) && !empty($params['membership_type_id'][1])) {
```
Comment (JoeMurray) [inline CRM/Member/Form/Membership.php]: I don't like the 1's here as a coding standard matter. Could we do something better?

File: CRM/Member/BAO/Membership.php (hunk 4 above)
Comment (monishdeb) [inline CRM/Member/BAO/Membership.php]: Actually, those two are reserved name, already used in Contribution BAO in various places. Should I change there too to ```partial_amount_to_pay```?

File: CRM/Member/Form/Membership.php (hunk 5 above)
Comment (monishdeb) [inline CRM/Member/Form/Membership.php]: ok will check.

File: CRM/Contribute/BAO/Contribution.php (hunk 2 above)
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Could you update the spelling of aginst to against? Thx

File: CRM/Member/BAO/Membership.php (hunk 4 above)
Comment (JoeMurray) [inline CRM/Member/BAO/Membership.php]: Yes please. Put in separate PR as cleanup.

Comment (monishdeb) [inline CRM/Member/BAO/Membership.php]: Done. new PR https://github.com/civicrm/civicrm-core/pull/10433

**Comment by monishdeb** [comment]
depends on https://github.com/civicrm/civicrm-core/pull/10433

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by colemanw** [comment]
@monishdeb can you please rebase this?

**Comment by monishdeb** [comment]
@colemanw done

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by monishdeb** [comment]
ping @eileenmcnaughton @jitendrapurohit 

This fix allows us to store correct financial entries when a partial payment for a membership is completed or made. Also, include the fix for remaining issue raised by @pradpnayak on https://github.com/civicrm/civicrm-core/pull/10981 , encountered during QA

Can you please review this PR?

**Comment by eileenmcnaughton** [comment]
I think we might need to break this out into multiple PRs as there is a lot going on - but I'll try to put some time aside to work through it & discuss with you

File: CRM/Contribute/BAO/Contribution.php [hunk 6]
```diff
@@ -3804,6 +3804,15 @@ public static function recordAdditionalPayment($contributionId, $trxnsData, $pay
       $participantId = CRM_Core_DAO::getFieldValue('CRM_Event_DAO_ParticipantPayment', $contributionId, 'participant_id', 'contribution_id');
     }
 
+    // fetch membership IDs registered against given contribution ID
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: So, I feel like the changes in this file, along with their test, should stand alone as a discrete PR & be fairly easy to get merged. My worry is that this seems to duplicate the code called by completeOrder, but do a bit less. Extract & share?

File: CRM/Financial/BAO/FinancialItem.php [hunk 7]
```diff
@@ -109,7 +109,9 @@ public static function add($lineItem, $contribution, $taxTrxnID = FALSE, $trxnId
     }
     else {
       $accountRelName = 'Income Account is';
-      if (property_exists($contribution, 'revenue_recognition_date') && !CRM_Utils_System::isNull($contribution->revenue_recognition_date)) {
+      if (property_exists($contribution, 'revenue_recognition_date') &&
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/FinancialItem.php]: this is a code smell. If it's necessary to fix an e-notice for tests sake I think we should define revenue_recognition_date as a BAO property (if not we can ignore for now :-)

**Comment by eileenmcnaughton** [comment]
I've only looked at the top 2 files so far (will set aside more time later) but the changes in both of them look like things that could be split off into small PRs that would be fairly easy to get merged

File: CRM/Contribute/BAO/Contribution.php (hunk 6 above)
Comment (monishdeb) [reply CRM/Contribute/BAO/Contribution.php]: yup will extract this piece of code along with related UT on a separate PR

**Comment by monishdeb** [comment]
@eileenmcnaughton I have moved portion of the patch into two separate PRs https://github.com/civicrm/civicrm-core/pull/10999 and #11000 

The related UT is expected to fail due to obvious reason. Should I further extract the code out of it?

**Comment by monishdeb** [comment]
Closing in favor of https://github.com/civicrm/civicrm-core/pull/11212
